targets/TARGET_NUVOTON/TARGET_M2351/TARGET_M23_NS/TARGET_TFM/nu_tfm_ns_lock_sup.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2019, Nuvoton Technology Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(DOMAIN_NS) && (DOMAIN_NS == 1L) && (TFM_LVL > 0)
+
+#include <stdint.h>
+#include "mbed_critical.h"
+#include "mbed_error.h"
+#include "nu_tfm_ns_lock_sup.h"
+#include "cmsis_os2.h"
+
+/* Enforce locked entry into TFM for platform-specific NSC calls
+ *
+ * On M2351, some spaces like SYS/CLK are hard-wired to secure and cannot change.
+ * To access these spaces from non-secure world, we must provide platform-specific NSC
+ * functions. With TFM introduced, we must synchronize NSC calls into TFM to keep TFM
+ * in sync instead of straight NSC calls. To achieve this goal, we go with the following
+ * approach:
+ *
+ * 1. Like standard PSA APIs, enforce locked entry through tfm_ns_lock_dispatch().
+ * 2. Run platform-specific secure functions in default secure partition, in which these
+ *    SYS/CLK spaces must have configured to be accessible.
+ */
+
+int32_t nu_tfm_ns_lock_dispatch(veneer_fn fn,
```

Comment:
What is the need for this Nuvoton specific API and when will it be called?
Currently, https://github.com/ARMmbed/mbed-os/blob/4c04771c5d3a8d730d41d7fdbc9859ffe7c12148/components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_psa_ns_api.c uses `tfm_ns_lock_dispatch` for dispatching secure calls to TF-M.

Reply:
`tfm_ns_lock_dispatch()` doesn't support secure call in pre-rtos or interrupt-disabled context, which is required to call platform-specific secure functions on M2351. In updated TF-M repo, `tfm_ns_lock_dispatch()` is renamed to `tfm_ns_interface_dispatch()` and gets weak https://github.com/ARMmbed/mbed-os/pull/10959#issuecomment-514446186. The `nu_tfm_ns_lock_dispatch(...)` is temporary for development. After new TF-M NS interface is integrated into mbed-os, `tfm_ns_interface_dispatch()` will be overridden in M2351 directory to try supporting secure call in pre-rtos or interrupt-disabled context.

Reply:
@devran01 Remove `nu_tfm_ns_lock_dispatch(...)`. Make `tfm_ns_lock_dispatch()` weak and override it to support secure call in unusual conditions like pre-rtos and interrupt-disabled.